File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-8.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        set_chars: set[str] = set()
        left: int = 0
        longest: int = 0

        for right in range(len(s)):
            while s[right] in set_chars:
                set_chars.remove(s[left])
                left += 1
            
            set_chars.add(s[right])
            longest = max(longest, right-left+1)

        return longest

File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-8.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    private int res = Integer.MIN_VALUE;
    public int maxPathSum(TreeNode root) {
        dfs(root);
        return res;
    }

    private int dfs(TreeNode node) {
        if (node == null) return 0;

        int leftMax = Math.max(dfs(node.left), 0);
        int rightMax = Math.max(dfs(node.right), 0);

        res = Math.max(res, node.val + leftMax + rightMax);

        return node.val + Math.max(leftMax, rightMax);
    }
}
